lib/src/main/kotlin/org/didcommx/didcomm/protocols/routing/Forward.kt

```diff
@@ -0,0 +1,231 @@
+            //      DID Doc services (it SHOULD NOT be as per spec but ...)
+            val errMsg = (
+                "mediator '${res.last().serviceEndpoint}' defines alternative" +
+                    " endpoint '$serviceUri' recursively"
+                )
+
+            if (didRecursion) {
+                throw NotImplementedError(errMsg)
+            } else {
+                throw DIDCommServiceException(res.last().serviceEndpoint, errMsg)
+            }
+        }
+
+        // TODO check not only first item in mediator services list
+        //      (e.g. first one may use alternative endpoint but second - URI)
+
+        // resolve until final URI is reached
+        val mediatorDidService = findDIDCommService(didDocResolver, mediatorDid)
+            ?: throw DIDCommServiceException(
+                mediatorDid, "mediator '$mediatorDid' service doc not found"
+            )
+
+        serviceUri = mediatorDidService.serviceEndpoint
+        res.add(0, mediatorDidService)
+    }
+
+    return res
+}
+
+fun wrapInForward(
+    packedMsg: Map<String, Any>,
+    to: String,
+    keySelector: SenderKeySelector,
+    encAlgAnon: AnonCryptAlg? = null,
+    routingKeys: List<String>? = null,
+    headers: Map<String, Any?>? = null
+): WrapInForwardResult? {
+    // means forward protocol is not needed
+    if (routingKeys == null)
+        return null
+
+    // TODO
+    //  - headers validation against ForwardMessage
+    //  - logging
+    //  - id generator as an argument
+
+    var fwdMsg: Message? = null
+    var forwardedMsg = packedMsg
+    var encryptedResult: EncryptResult? = null
+
+    val tos = routingKeys.asReversed()
+    val nexts = (routingKeys.drop(1) + to).asReversed()
+
+    // wrap forward msgs in reversed order so the message to final
+    // recipient 'to' will be the innermost one
+    for ((_to, _next) in tos.zip(nexts)) {
+        val fwdAttach = Attachment.builder(
+            didcommIdGeneratorDefault(), Attachment.Data.Json(forwardedMsg)
+        ).build()
+        // TODO ??? .mediaType("application/json")
+
+        val fwdMsgBuilder = Message.builder(
+            didcommIdGeneratorDefault(),
+            mapOf("next" to _next),
+            DIDCommMessageProtocolTypes.Forward.typ
+        ).attachments(listOf(fwdAttach))
+
+        headers?.forEach { (name, value) ->
+            fwdMsgBuilder.customHeader(name, value)
+        }
+
+        fwdMsg = fwdMsgBuilder.build()
+
+        // TODO improve: do not rebuild each time 'to' is changed
+        val packParamsBuilder = PackEncryptedParams.Builder(fwdMsg, _to)
+
+        if (encAlgAnon != null)
+            packParamsBuilder.encAlgAnon(encAlgAnon)
+
+        encryptedResult = encrypt(
+            packParamsBuilder.build(), fwdMsg.toString(), keySelector
+        ).first
+
+        forwardedMsg = JSONObjectUtils.parse(encryptedResult.packedMessage)
+    }
+
+    encryptedResult = encryptedResult!!
```

Comment:
Not good to call !!. If null can be there, then handle it and throw exception if needed.

Reply:
improved, thanks